File: app/src/main/java/com/oseemasuaku/codedutravail/data/repositories/ChapterRepositoryImpl.kt
package com.oseemasuaku.codedutravail.data.repositories

import com.oseemasuaku.codedutravail.data.daos.ChapterDao
import com.oseemasuaku.codedutravail.data.entities.Chapter
import com.oseemasuaku.codedutravail.domain.entities.ChapterData
import com.oseemasuaku.codedutravail.domain.repositories.ChapterRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

class ChapterRepositoryImpl(
    private val chapterDao: ChapterDao
): ChapterRepository {
    override suspend fun insertAll(vararg chapters: ChapterData) {
        val dbChapters = chapters.map {
            Chapter(number = it.number, text = it.text, title =  it.titleNumber, articles = it.articles)
        }
        chapterDao.insertAll(*dbChapters.toTypedArray())
    }

    override fun streamByTitle(titleId: Int): Flow<List<ChapterData>> {
        return chapterDao.streamByTitle(titleId).map { data ->
            data.map {
                ChapterData(it.number, it.text, it.title, it.articles)
            }
        }
    }
}

File: lib/data/repositories/chapter_repository_impl.dart
import 'package:codedutravail/core/data/database.dart';
import 'package:codedutravail/data/tables/chapters.dart';
import 'package:codedutravail/domain/entities/chapter.dart';
import 'package:codedutravail/domain/repositories/chapter_repository.dart';
import 'package:drift/drift.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';

part 'chapter_repository_impl.g.dart';

@Riverpod(keepAlive: true)
ChapterRepository chapterRepository(Ref ref) {
  final db = ref.watch(dbProvider);

  return ChapterRepositoryImpl(db);
}

@DriftAccessor(tables: [Chapters])
class ChapterRepositoryImpl extends DatabaseAccessor<AppDatabase> with _$ChapterRepositoryImplMixin, ChapterRepository {
  ChapterRepositoryImpl(super.attachedDatabase);

  @override
  Future<void> insertAll(List<ChapterEntity> entries) {
    final rows = entries.map(
      (entry) => ChaptersCompanion.insert(
        articles: entry.articles,
        value: entry.text,
        number: entry.number,
        titleId: entry.titleNumber,
      ),
    );
    return batch((batch) {
      batch.insertAllOnConflictUpdate(chapters, rows);
    });
  }

  @override
  Stream<List<ChapterEntity>> streamByTitle(int titleId) {
    return (select(chapters)..where((t) => t.titleId.equals(titleId))).watch().asyncMap(
      (values) =>
          values
              .map(
                (value) => ChapterEntity(
                  number: value.number,
                  text: value.value,
                  titleNumber: value.titleId,
                  articles: value.articles,
                ),
              )
              .toList(),
    );
  }
}
